src/passes/OptimizeCasts.cpp

```diff
@@ -102,6 +121,268 @@ namespace wasm {
 
 namespace {
 
+// Contains information about a RefCast we want to move to a target LocalGet.
+struct RefCastInfo {
+  LocalGet* target = nullptr;
+  RefCast* bestCast = nullptr;
+};
+
+// Contains information about a RefAs we want to move to a target LocalGet.
+// Currently only RefAsNonNull will be moved.
+struct RefAsInfo {
+  LocalGet* target = nullptr;
+  RefAs* bestCast = nullptr;
+};
+
+// Find a cast to move earlier to another local.get. More refined subtypes are
+// chosen over less refined ones.
+struct EarlyCastFinder
+  : public LinearExecutionWalker<EarlyCastFinder,
+                                 UnifiedExpressionVisitor<EarlyCastFinder>> {
+  PassOptions options;
+  size_t numLocals;
+
+  // Tracks the current earliest local.get that we can move a cast to without
```

Comment:
Please start with `For each local index,` - otherwise it's not obvious what these vectors are indexed on, I think.

Reply:
Done.